README.md
LIB-PM-Cell_Toolkit
===================

Cell toolkit is a library for OS/2 PM designed to simplify implementation of two important parts of user interface (UI). Split views and Toolbars 

LICENSE
===============
* BSD 2 Clauses

COMPILE TOOLS
===============
* 

AUTHORS
===============
* Sergey I. Yevtushenko

LINKS
===============
* http://es.os2.ru/cell.html


history.txt
/*
** Module   :HISTORY.TXT
** Abstract :
**
** Copyright (C) Sergey I. Yevtushenko
**
** Log: Thu  29/04/1999 Created
**      Wed  25/10/2000 Updated for version 0.7b
*/

+ - added
* - fixed
- - dropped

v0.7b
    + Support for swapping cells via SetSplitType()
    + Support for "rotating" splitbar via RMB click on splitbar
    + Support for hiding cells at run time
    * Minor bug fixes and enchancements

v0.6e
    * Some samples and makefile cleaned up



samples/entry.cpp
/*
** Module   :SIMPLE.C
** Abstract :
**
** Copyright (C) Sergey I. Yevtushenko
** Log: Sun  14/02/1998     Created
**
*/

#define INCL_WIN
#define INCL_NLS
#define INCL_DOS
#define INCL_GPI

#include <os2.h>
#include <multibar.h>
#include <cell.h>
#include <string.h>

/* Local procedures */

MRESULT EXPENTRY MainClientProc(HWND hwnd, ULONG msg, MPARAM mp1, MPARAM mp2);
MRESULT EXPENTRY LEProc(HWND hwnd, ULONG msg, MPARAM mp1, MPARAM mp2);

/* Static Variables */
#define ID_FINDEF   1060
#define ID_FINDPB   1061
#define ID_PRXCELL  1062
#define ID_MLE2     1063


CellDef cdLeftPane =
{
    CELL_WINDOW,
    WC_MLE,
    "SampleText",
    MLS_BORDER | WS_VISIBLE,
    ID_MLE2
};

CellDef findEf =
{
    CELL_WINDOW,
    WC_ENTRYFIELD,
    NULL,
    WS_VISIBLE | ES_ANY | ES_MARGIN,
    ID_FINDEF,
    0,0,
    LEProc,
    0,0
};

CellDef findPb =
{
    CELL_WINDOW,
    WC_BUTTON,
    "Find",
    WS_VISIBLE,
    ID_FINDPB
};

CellDef cdRightPane =
{
    CELL_VSPLIT | CELL_SIZE2 | CELL_SPLITBAR | CELL_FIXED,
    0,
    NULL,
    WS_VISIBLE,
    ID_PRXCELL,
    &findEf,
    &findPb,
    NULL, NULL,
    60
};

CellDef mainClient =
{
    CELL_HSPLIT | CELL_SIZE2 | CELL_FIXED,
    0,
    "Simple",
    FCF_TITLEBAR | FCF_SYSMENU  | FCF_MENU |
    FCF_MINMAX   | FCF_TASKLIST | FCF_SIZEBORDER |
    0,
    MAIN_FRAME,
    &cdLeftPane,
    &cdRightPane,
    0,
    MainClientProc,
    20
};

ULONG mainItems[]=
{
    IDB_FILENEW ,
    IDB_FILEOPEN,
    IDB_FILESAVE,
    IDB_FILSAVAS,
    TB_SEPARATOR,
    IDB_EXIT    ,
    TB_SEPARATOR,
    IDB_EDITCOPY,
    IDB_EDITCUT ,
    IDB_EDITPAST,
    IDB_EDITUNDO,
    TB_SEPARATOR,
    IDB_EDITFIND,
    IDB_EDITFNNX,
    IDB_EDITREPL,
    TB_SEPARATOR,
    IDB_HELP    ,
    IDB_ABOUT   ,
    0
};

TbDef mainTb =
{
    TB_ATTACHED_TP /*| TB_BUBBLE*/,
    ID_TOOLBAR,
    mainItems
};

MRESULT EXPENTRY LEProc(HWND hwnd, ULONG msg, MPARAM mp1, MPARAM mp2)
{
    WindowCellCtlData *pWCtlData = 0;

    pWCtlData = (WindowCellCtlData *)WinQueryWindowULong(hwnd, QWL_USER);

    switch(msg)
    {
        case WM_BUTTON1DOWN:
            {
                WinPostMsg(hwnd, WM_BUTTON1DBLCLK, mp1, mp2);
            }
            break;
    }
    if(pWCtlData)
        return pWCtlData->pOldProc(hwnd, msg, mp1, mp2);

    return WinDefWindowProc(hwnd, msg, mp1, mp2);
}

MRESULT EXPENTRY MainClientProc(HWND hwnd, ULONG msg, MPARAM mp1, MPARAM mp2)
{
    WindowCellCtlData *pWCtlData = 0;
    static iTimerStarted = 0;
/*
    if(!iTimerStarted)
    {
        APIRET rc = WinStartTimer(0, hwnd, 1, 1000);
        iTimerStarted = 0;
    }
*/
    pWCtlData = (WindowCellCtlData *)WinQueryWindowULong(hwnd, QWL_USER);

    switch(msg)
    {
        case WM_TIMER:
            WinSetWindowPos(WinQueryWindow(hwnd, QW_PARENT),
                        	HWND_TOP, 0,0,0,0, SWP_ZORDER);
            break;

        case WM_COMMAND:
            switch(SHORT1FROMMP(mp1))
            {
                case IDB_EXIT:
					WinPostMsg(hwnd, WM_QUIT, 0L, 0L);
                    return ((MRESULT) NULL);
            }
            break;

        case WM_CLOSE:
			WinPostMsg(hwnd, WM_QUIT, 0L, 0L);
			return ((MRESULT) NULL);
    }
    if(pWCtlData)
        return pWCtlData->pOldProc(hwnd, msg, mp1, mp2);

    return WinDefWindowProc(hwnd, msg, mp1, mp2);
}

INT main(VOID)
{
	HAB hab;
	HMQ hmq;
	QMSG qmsg;
    HWND hwndFrame;
    HWND hwndTb;
    HWND hwndTmp;
    LONG lColor = CLR_PALEGRAY;
    static char cFontMy[]="8.Helv";
    SWP swp;

    hab = WinInitialize(0);

    if(!hab)
    {
        return -1;
    }

    hmq = WinCreateMsgQueue(hab, 0);

    if(!hmq)
    {
        WinTerminate(hab);
        return -2;
    }

    ToolkitInit(hab);

    WinQueryWindowPos(HWND_DESKTOP, &swp);

    hwndFrame = CreateCell(&mainClient, HWND_DESKTOP, 0);
    if(hwndFrame)
    {
        WinSetWindowPos(hwndFrame, NULLHANDLE,
        	            swp.x + swp.cx / 8,
            	        swp.y + swp.cy / 8,
                	    (swp.cx / 4) * 3,
                    	(swp.cy / 4) * 3,
	                    SWP_ACTIVATE | SWP_MOVE | SWP_SIZE | SWP_SHOW);


        CreateToolbar(hwndFrame, &mainTb);

        /* Set MLE color */

        hwndTmp = CellWindowFromID(hwndFrame, ID_MLE2);

        WinSetPresParam(hwndTmp,
                        PP_BACKGROUNDCOLORINDEX,
                        sizeof(lColor),
                        &lColor);

        WinSetPresParam(hwndFrame,
                        PP_FONTNAMESIZE,
                        sizeof(cFontMy),
                        cFontMy);

        WinStartTimer(0, WinWindowFromID(hwndFrame, FID_CLIENT), 1, 1000);

        while (WinGetMsg(hab, &qmsg, 0, 0, 0))
        	WinDispatchMsg(hab, &qmsg);

        WinDestroyWindow(hwndFrame);
    }

	WinDestroyMsgQueue(hmq);
	WinTerminate(hab);

    return 0;
}




samples/multibar.h
/*
** Module   :MULTIBAR.H
** Abstract :Constants for MULTIBAR.C
**
** Copyright (C) Sergey I. Yevtushenko
** Log: Sun  08/02/98   Refined
*/

#ifndef __MULTIBAR_H
#define __MULTIBAR_H

#define MAIN_FRAME    255
#define ID_FILEMENU   256
#define ID_EDITMENU   258

#define ID_TOOLBAR    1045
#define ID_LPANE      1046
#define ID_RPANE      1047
#define ID_NOTEBOOK   1045
#define ID_STATUS     1050
#define ID_LIST       1051
#define ID_MLE        1052
#define ID_TOP        1053

#define IDB_FILENEW   301
#define IDB_FILEOPEN  302
#define IDB_FILESAVE  303
#define IDB_FILSAVAS  304
#define IDB_EXIT      305
#define IDB_EDITCOPY  306
#define IDB_EDITCUT   307
#define IDB_EDITFIND  308
#define IDB_EDITFNNX  309
#define IDB_EDITPAST  310
#define IDB_EDITREPL  311
#define IDB_EDITUNDO  312
#define IDB_HELP      313
#define IDB_ABOUT     314

#endif



samples/simple.cpp
/*
** Module   :SIMPLE.C
** Abstract :
**
** Copyright (C) Sergey I. Yevtushenko
** Log: Sun  14/02/1998     Created
**
*/

#define INCL_WIN
#define INCL_NLS
#define INCL_DOS
#define INCL_GPI

#include <os2.h>
#include <multibar.h>
#include <cell.h>
#include <string.h>

/* Local procedures */

MRESULT EXPENTRY MainClientProc(HWND hwnd, ULONG msg, MPARAM mp1, MPARAM mp2);

/* Static Variables */

CellDef cdLeftPane =
{
    CELL_WINDOW,
    WC_LISTBOX,
    "List",
    LS_NOADJUSTPOS | WS_VISIBLE,
    ID_LIST
};

CellDef cdRightPane =
{
    CELL_WINDOW,
    WC_MLE,
    "SampleText",
    MLS_BORDER | WS_VISIBLE,
    ID_MLE
};

CellDef mainClient =
{
    CELL_VSPLIT | CELL_SPLITBAR | CELL_SPLIT30x70,
    0,
    "Simple",
    FCF_TITLEBAR | FCF_SYSMENU  | FCF_MENU |
    FCF_MINMAX   | FCF_TASKLIST | FCF_SIZEBORDER |
    0,
    MAIN_FRAME,
    &cdLeftPane,
    &cdRightPane,
    0,
    MainClientProc
};

ULONG mainItems[]=
{
    IDB_FILENEW ,
    IDB_FILEOPEN,
    IDB_FILESAVE,
    IDB_FILSAVAS,
    TB_SEPARATOR,
    IDB_EXIT    ,
    TB_SEPARATOR,
    IDB_EDITCOPY,
    IDB_EDITCUT ,
    IDB_EDITPAST,
    IDB_EDITUNDO,
    TB_SEPARATOR,
    IDB_EDITFIND,
    IDB_EDITFNNX,
    IDB_EDITREPL,
    TB_SEPARATOR,
    IDB_HELP    ,
    IDB_ABOUT   ,
    0
};

TbDef mainTb =
{
    TB_ATTACHED_TP | TB_BUBBLE,
    ID_TOOLBAR,
    mainItems
};

MRESULT EXPENTRY MainClientProc(HWND hwnd, ULONG msg, MPARAM mp1, MPARAM mp2)
{
    WindowCellCtlData *pWCtlData = 0;

    pWCtlData = (WindowCellCtlData *)WinQueryWindowULong(hwnd, QWL_USER);

    switch(msg)
    {
        case WM_COMMAND:
            switch(SHORT1FROMMP(mp1))
            {
                case IDB_EXIT:
					WinPostMsg(hwnd, WM_QUIT, 0L, 0L);
                    return ((MRESULT) NULL);
            }
            break;

        case WM_CLOSE:
			WinPostMsg(hwnd, WM_QUIT, 0L, 0L);
			return ((MRESULT) NULL);
    }
    if(pWCtlData)
        return pWCtlData->pOldProc(hwnd, msg, mp1, mp2);

    return WinDefWindowProc(hwnd, msg, mp1, mp2);
}

INT main(VOID)
{
	HAB hab;
	HMQ hmq;
	QMSG qmsg;
    HWND hwndFrame;
    HWND hwndTb;
    HWND hwndTmp;
    LONG lColor = CLR_PALEGRAY;
    SWP swp;

    hab = WinInitialize(0);

    if(!hab)
    {
        return -1;
    }

    hmq = WinCreateMsgQueue(hab, 0);

    if(!hmq)
    {
        WinTerminate(hab);
        return -2;
    }

    ToolkitInit(hab);

    WinQueryWindowPos(HWND_DESKTOP, &swp);

    hwndFrame = CreateCell(&mainClient, HWND_DESKTOP, 0);
    if(hwndFrame)
    {
        HWND hwndList;

        hwndList = CellWindowFromID(hwndFrame, ID_LPANE);

        if(hwndList)
        {
            CHAR cText[256];
            static char cListItem[] = "List item ";
            int i;

            strcpy(cText, cListItem);

            for(i = 0; i < 15; i++)
            {
                GenResIDStr(cText + sizeof(cListItem) - 1, i);
                WinSendMsg(hwndList,
                           LM_INSERTITEM,
                           MPFROMSHORT(LIT_END),
                           MPFROMP(cText));
            }
        }

        WinSetWindowPos(hwndFrame, NULLHANDLE,
        	            swp.x + swp.cx / 8,
            	        swp.y + swp.cy / 8,
                	    (swp.cx / 4) * 3,
                    	(swp.cy / 4) * 3,
	                    SWP_ACTIVATE | SWP_MOVE | SWP_SIZE | SWP_SHOW);


        CreateToolbar(hwndFrame, &mainTb);

        /* Set MLE color */

        hwndTmp = CellWindowFromID(hwndFrame, ID_MLE);

        WinSendMsg(hwndTmp,
                   MLM_SETBACKCOLOR,
                   MPFROMLONG(CLR_PALEGRAY),
                   MPFROMLONG(MLE_INDEX));

        /* Set list color */

        hwndTmp = CellWindowFromID(hwndFrame, ID_LIST);

        WinSetPresParam(hwndTmp,
                        PP_BACKGROUNDCOLORINDEX,
                        sizeof(lColor),
                        &lColor);

        while (WinGetMsg(hab, &qmsg, 0, 0, 0))
        	WinDispatchMsg(hab, &qmsg);

        WinDestroyWindow(hwndFrame);
    }

	WinDestroyMsgQueue(hmq);
	WinTerminate(hab);

    return 0;
}




samples/status.cpp
/*
** Module   :MULTIBAR.C
** Abstract :Test and sample application for Cell toolkit procedures.
**
** Copyright (C) Sergey I. Yevtushenko
** Log: Sun  08/02/98   Refined
**
*/

#define INCL_WIN
#define INCL_NLS
#define INCL_DOS
#define INCL_GPI

#include <os2.h>
#include <multibar.h>
#include <cell.h>

/* Local procedures */

MRESULT EXPENTRY MainClientProc(HWND hwnd, ULONG msg, MPARAM mp1, MPARAM mp2);

/* Static Variables */

CellDef cdLeftPane =
{
    CELL_WINDOW,
    WC_LISTBOX,
    "List",
    LS_NOADJUSTPOS | WS_VISIBLE,
    ID_LIST
};

CellDef cdRightPane =
{
    CELL_WINDOW,
    WC_MLE,
    "SampleText",
    MLS_BORDER | WS_VISIBLE,
    ID_MLE
};

CellDef rPanel =
{
    CELL_VSPLIT | CELL_SPLITBAR | CELL_SPLIT30x70,
    0,
    "Subwindow",
    WS_VISIBLE,
    ID_TOP,
    &cdLeftPane,
    &cdRightPane,
};

CellDef Panel1 =
{
    CELL_WINDOW,
    "StatusLine",
    "",
    WS_VISIBLE,
    ID_STATUS
};

CellDef mainClient =
{
    CELL_HSPLIT | CELL_FIXED | CELL_SIZE1,
    0,
    "Status Line Sample",
    FCF_TITLEBAR | FCF_SYSMENU  | FCF_MENU |
    FCF_MINMAX   | FCF_TASKLIST | FCF_SIZEBORDER |
    0,
    MAIN_FRAME,
    &Panel1,
    &rPanel,
    0,                  //Frame subclass proc
    MainClientProc,     //Client subclass proc
    20                  //Status line hight
};

ULONG mainItems[]=
{
    IDB_FILENEW ,
    IDB_FILEOPEN,
    IDB_FILESAVE,
    IDB_FILSAVAS,
    TB_SEPARATOR,
    IDB_EXIT    ,
    TB_SEPARATOR,
    IDB_EDITCOPY,
    IDB_EDITCUT ,
    IDB_EDITFIND,
    IDB_EDITFNNX,
    IDB_EDITPAST,
    IDB_EDITREPL,
    IDB_EDITUNDO,
    TB_SEPARATOR,
    IDB_HELP    ,
    IDB_ABOUT   ,
    0
};

TbDef mainTb =
{
    TB_VERTICAL | TB_ATTACHED_TP | TB_BUBBLE,
    ID_TOOLBAR,
    mainItems
};

ULONG OKMsgBox(PSZ pszText)
{
    return WinMessageBox(HWND_DESKTOP,
                         HWND_DESKTOP,
                         pszText,
                         "Cell Demo",
                         0,
                         MB_OK | MB_INFORMATION | MB_APPLMODAL);
}

MRESULT EXPENTRY MainClientProc(HWND hwnd, ULONG msg, MPARAM mp1, MPARAM mp2)
{
    WindowCellCtlData *pWCtlData = 0;

    pWCtlData = (WindowCellCtlData *)WinQueryWindowULong(hwnd, QWL_USER);

    switch(msg)
    {
        case WM_COMMAND:
            switch(SHORT1FROMMP(mp1))
            {
                case IDB_EXIT:
					WinPostMsg(hwnd, WM_QUIT, 0L, 0L);
                    return ((MRESULT) NULL);
            }
            break;

        case WM_CLOSE:
			WinPostMsg(hwnd, WM_QUIT, 0L, 0L);
			return ((MRESULT) NULL);
    }
    if(pWCtlData)
        return pWCtlData->pOldProc(hwnd, msg, mp1, mp2);

    return WinDefWindowProc(hwnd, msg, mp1, mp2);
}

MRESULT EXPENTRY StatusLineProc(HWND hwnd, ULONG msg, MPARAM mp1, MPARAM mp2)
{
    switch(msg)
    {
        case WM_PAINT:
            {
                HPS hpsPaint;
                RECTL rclPaint;
                POINTL ptlWork;

                hpsPaint = WinBeginPaint(hwnd, 0, 0);
                WinQueryWindowRect(hwnd, &rclPaint);

                WinFillRect(hpsPaint, &rclPaint, CLR_PALEGRAY);

                GpiSetColor(hpsPaint, CLR_BLACK);

                ptlWork.x = rclPaint.xLeft      ;
                ptlWork.y = rclPaint.yBottom    ;
                GpiMove(hpsPaint, &ptlWork);
                ptlWork.y = rclPaint.yTop    - 2;
                GpiLine(hpsPaint, &ptlWork);
                ptlWork.x = rclPaint.xRight  - 1;
                GpiLine(hpsPaint, &ptlWork);

                GpiSetColor(hpsPaint,CLR_WHITE);

                ptlWork.y = rclPaint.yBottom    ;
                GpiLine(hpsPaint, &ptlWork);
                ptlWork.x = rclPaint.xLeft      ;
                GpiLine(hpsPaint, &ptlWork);

                rclPaint.yTop    -= 3;
                rclPaint.yBottom += 1;
                rclPaint.xRight  -= 2;
                rclPaint.xLeft   += 1;

                WinDrawText(hpsPaint, -1, "Status message", &rclPaint,
                            CLR_BLACK,
                            0,
                            DT_LEFT | DT_VCENTER);

                WinEndPaint(hpsPaint);
            }
            return MRFROMSHORT(FALSE);
    }
    return WinDefWindowProc(hwnd, msg, mp1, mp2);
}

INT main(VOID)
{
	HAB hab;
	HMQ hmq;
	QMSG qmsg;
    HWND hwndFrame;
    HWND hwndTb;
    HWND hwndTmp;
    LONG lColor = CLR_PALEGRAY;
    SWP swp;
    static char cFontMy[]="8.Helv";

    hab = WinInitialize(0);

    if(!hab)
    {
        return -1;
    }

    hmq = WinCreateMsgQueue(hab, 0);

    if(!hmq)
    {
        WinTerminate(hab);
        return -2;
    }

    ToolkitInit(hab);

    WinRegisterClass(hab,
                     "StatusLine",
                     StatusLineProc,
                     CS_SIZEREDRAW,
                     sizeof(ULONG));


    WinQueryWindowPos(HWND_DESKTOP, &swp);

    hwndFrame = CreateCell(&mainClient, HWND_DESKTOP, 0);

    if(hwndFrame)
    {
        HWND hwndSubframe = CellWindowFromID(hwndFrame, ID_LPANE);

        WinSetWindowPos(hwndFrame, NULLHANDLE,
        	            swp.x + swp.cx / 8,
            	        swp.y + swp.cy / 8,
                	    (swp.cx / 4) * 3,
                    	(swp.cy / 4) * 3,
	                    SWP_ACTIVATE | SWP_MOVE | SWP_SIZE | SWP_SHOW);


        CreateToolbar(hwndFrame, &mainTb);

        /* Set status line font */

        hwndTmp = CellWindowFromID(hwndFrame, ID_STATUS);

        WinSetPresParam(hwndTmp,
    	                PP_FONTNAMESIZE,
            	        sizeof(cFontMy),
                        cFontMy);

        /* Set MLE color */

        hwndTmp = CellWindowFromID(hwndFrame, ID_MLE);

        WinSendMsg(hwndTmp,
    	           MLM_SETBACKCOLOR,
                   MPFROMLONG(CLR_PALEGRAY),
                   MPFROMLONG(MLE_INDEX));

        /* Set list color */

        hwndTmp = CellWindowFromID(hwndFrame, ID_LIST);

        WinSetPresParam(hwndTmp,
                        PP_BACKGROUNDCOLORINDEX,
                        sizeof(lColor),
                        &lColor);

        while (WinGetMsg(hab, &qmsg, 0, 0, 0))
        	WinDispatchMsg(hab, &qmsg);

        WinDestroyWindow(hwndFrame);
    }

	WinDestroyMsgQueue(hmq);
	WinTerminate(hab);

    return 0;
}




samples/multibar.cpp
/*
** Module   :MULTIBAR.C
** Abstract :Test and sample application for Cell toolkit procedures.
**
** Copyright (C) Sergey I. Yevtushenko
** Log: Sun  08/02/98   Refined
**
*/

#define INCL_WIN
#define INCL_NLS
#define INCL_DOS
#define INCL_GPI

#include <os2.h>
#include <multibar.h>
#include <cell.h>

/* Local procedures */

MRESULT EXPENTRY MainClientProc(HWND hwnd, ULONG msg, MPARAM mp1, MPARAM mp2);

/* Static Variables */

CellDef Panel0 =
{
    CELL_WINDOW,
    WC_BUTTON,
    "Button",
    BS_PUSHBUTTON | WS_VISIBLE,
    ID_RPANE
};

CellDef Panel1 =
{
    CELL_WINDOW,
    WC_STATIC,
    "Button",
    SS_TEXT | DT_CENTER | DT_VCENTER | WS_VISIBLE,
    ID_LPANE
};

CellDef uPanel =
{
    CELL_HSPLIT | CELL_SPLITBAR | CELL_FIXED | CELL_SIZE2,
    0,
    "",
    WS_VISIBLE,
    ID_RPANE,
    &Panel0,
    &Panel1,
    0,
    0,
    40
};

CellDef dPanel =
{
    CELL_WINDOW,
    WC_STATIC,
    "Button",
    SS_TEXT | DT_CENTER | DT_VCENTER | WS_VISIBLE,
    ID_LPANE,
    0,
    0
};

CellDef rPanel =
{
    CELL_VSPLIT | CELL_SPLITBAR | CELL_SPLIT30x70,
    0,
    "Subwindow",
    WS_VISIBLE,
    ID_RPANE,
    &uPanel,
    &dPanel
};

CellDef lPanel =
{
    CELL_VSPLIT | CELL_SPLITBAR | CELL_SPLIT70x30,
    0,
    "",
    WS_VISIBLE,
    ID_LPANE,
    &uPanel,
    &dPanel
};

CellDef mainClient =
{
    CELL_HSPLIT | CELL_SPLITBAR,
    0,
    "Multiple Toolbars Sample",
    FCF_TITLEBAR | FCF_SYSMENU  | FCF_MENU |
    FCF_MINMAX   | FCF_TASKLIST | FCF_SIZEBORDER |
    0,
    MAIN_FRAME,
    &lPanel,
    &rPanel,
    0,                  //Frame subclass proc
    MainClientProc      //Client subclass proc
};

ULONG mainItems[]=
{
    IDB_FILENEW ,
    IDB_FILEOPEN,
    IDB_FILESAVE,
    IDB_FILSAVAS,
    TB_SEPARATOR,
    IDB_EXIT    ,
    TB_SEPARATOR,
    IDB_EDITCOPY,
    IDB_EDITCUT ,
    IDB_EDITFIND,
    IDB_EDITFNNX,
    IDB_EDITPAST,
    IDB_EDITREPL,
    IDB_EDITUNDO,
    TB_SEPARATOR,
    IDB_HELP    ,
    IDB_ABOUT   ,
    0
};

ULONG mainItems1[]=
{
    IDB_FILENEW ,
    TB_SEPARATOR,
    IDB_FILEOPEN,
    IDB_FILESAVE,
    IDB_FILSAVAS,
    0
};

ULONG mainItems2[]=
{
    IDB_EDITCOPY,
    IDB_EDITCUT ,
    IDB_EDITPAST,
    TB_SEPARATOR,
    IDB_EDITFIND,
    IDB_EDITFNNX,
    IDB_EDITREPL,
    TB_SEPARATOR,
    IDB_EDITUNDO,
    0
};

ULONG mainItems3[]=
{
    IDB_HELP ,
    IDB_ABOUT,
    0
};

ULONG mainItems4[]=
{
    IDB_EXIT    ,
    TB_SEPARATOR,
    IDB_HELP    ,
    IDB_ABOUT   ,
    0
};


TbDef mainTb =
{
    TB_VERTICAL | TB_ATTACHED_LT | TB_BUBBLE,
    ID_TOOLBAR,
    mainItems
};

TbDef mainTb1 =
{
    TB_VERTICAL | TB_ATTACHED_RT | TB_BUBBLE,
    ID_TOOLBAR,
    mainItems1
};

TbDef mainTb2 =
{
    0           | TB_ATTACHED_TP | TB_BUBBLE,
    ID_TOOLBAR,
    mainItems2
};

TbDef mainTb3 =
{
    0           | TB_ATTACHED_BT | TB_BUBBLE,
    ID_TOOLBAR,
    mainItems3
};

TbDef mainTb4 =
{
    TB_VERTICAL | TB_FLOATING | TB_BUBBLE,
    ID_TOOLBAR,
    mainItems4
};

TbDef mainTb5 =
{
    0           | TB_FLOATING | TB_BUBBLE,
    ID_TOOLBAR,
    mainItems
};

ULONG OKMsgBox(PSZ pszText)
{
    return WinMessageBox(HWND_DESKTOP,
                         HWND_DESKTOP,
                         pszText,
                         "Cell Demo",
                         0,
                         MB_OK | MB_INFORMATION | MB_APPLMODAL);
}

MRESULT EXPENTRY MainClientProc(HWND hwnd, ULONG msg, MPARAM mp1, MPARAM mp2)
{
    WindowCellCtlData *pWCtlData = 0;

    pWCtlData = (WindowCellCtlData *)WinQueryWindowULong(hwnd, QWL_USER);

    switch(msg)
    {
        case WM_COMMAND:
            switch(SHORT1FROMMP(mp1))
            {
                case IDB_EXIT:
					WinPostMsg(hwnd, WM_QUIT, 0L, 0L);
                    return ((MRESULT) NULL);

                case ID_LPANE  :OKMsgBox("Button ID_LPANE pressed!"); break;
                case ID_RPANE  :OKMsgBox("Button ID_RPANE pressed!"); break;
                case MAIN_FRAME:OKMsgBox("Button MAIN_FRAME pressed!"); break;
                case ID_TOOLBAR:OKMsgBox("Button ID_TOOLBAR pressed!"); break;
            }
            break;

        case WM_CLOSE:
			WinPostMsg(hwnd, WM_QUIT, 0L, 0L);
			return ((MRESULT) NULL);
    }
    if(pWCtlData)
        return pWCtlData->pOldProc(hwnd, msg, mp1, mp2);

    return WinDefWindowProc(hwnd, msg, mp1, mp2);
}

INT main(VOID)
{
	HAB hab;
	HMQ hmq;
	QMSG qmsg;
    HWND hwndFrame;
    HWND hwndTb;
    SWP swp;

    hab = WinInitialize(0);

    if(!hab)
    {
        return -1;
    }

    hmq = WinCreateMsgQueue(hab, 0);

    if(!hmq)
    {
        WinTerminate(hab);
        return -2;
    }

    ToolkitInit(hab);

    WinQueryWindowPos(HWND_DESKTOP, &swp);

    hwndFrame = CreateCell(&mainClient, HWND_DESKTOP, 0);
    if(hwndFrame)
    {
        HWND hwndSubframe = CellWindowFromID(hwndFrame, ID_LPANE);

        WinSetWindowPos(hwndFrame, NULLHANDLE,
        	            swp.x + swp.cx / 8,
            	        swp.y + swp.cy / 8,
                	    (swp.cx / 4) * 3,
                    	(swp.cy / 4) * 3,
	                    SWP_ACTIVATE | SWP_MOVE | SWP_SIZE | SWP_SHOW);


        CreateToolbar(hwndFrame, &mainTb);
        CreateToolbar(hwndFrame, &mainTb1);
        CreateToolbar(hwndFrame, &mainTb2);
        CreateToolbar(hwndFrame, &mainTb3);
        CreateToolbar(hwndFrame, &mainTb4);

        if(hwndSubframe)
            CreateToolbar(hwndSubframe, &mainTb5);

        while (WinGetMsg(hab, &qmsg, 0, 0, 0))
        	WinDispatchMsg(hab, &qmsg);

        WinDestroyWindow(hwndFrame);
    }

	WinDestroyMsgQueue(hmq);
	WinTerminate(hab);

    return 0;
}




libcell/cell.h
/*
** Module   :CELL.H
** Abstract :Cell Toolkit header file
**
** Copyright (C) Sergey I. Yevtushenko
** Log: Sun  08/02/98   Created
**      Wed  25/10/2000 Updated to version 0.7b
*/

#ifndef __CELL_H
#define __CELL_H

/* Constants */

#define TK_VERSION      "0.7b"  /* Toolkit version */
#define CELL_WINDOW     0x0000   /* Cell is window */
#define CELL_VSPLIT     0x0001   /* Cell is vertically splitted view */
#define CELL_HSPLIT     0x0002   /* Cell is horizontally splitted view */
#define CELL_SPLITBAR   0x0004   /* Call has a splitbar */
#define CELL_FIXED      0x0008   /* Views can't be sized */
#define CELL_SIZE1      0x0010   /* */
#define CELL_SIZE2      0x0020
#define CELL_HIDE_1     0x0040   /* Cell 1 is hidden */
#define CELL_HIDE_2     0x0080   /* Cell 2 is hidden */
#define CELL_HIDE       0x00C0   /* Cell 1 or cell 2 is hidden */
#define CELL_SPLIT_MASK 0x003F
#define CELL_SWAP       0x1000   /* Cells are swapped */
#define CELL_SPLIT10x90 0x0100   /* Sizes of panels related as 10% and 90% */
#define CELL_SPLIT20x80 0x0200   /* Sizes of panels related as 20% and 80% */
#define CELL_SPLIT30x70 0x0300   /* Sizes of panels related as 30% and 70% */
#define CELL_SPLIT40x60 0x0400   /* Sizes of panels related as 40% and 60% */
#define CELL_SPLIT50x50 0x0500   /* Sizes of panels related as 50% and 50% */
#define CELL_SPLIT60x40 0x0600   /* Sizes of panels related as 60% and 40% */
#define CELL_SPLIT70x30 0x0700   /* Sizes of panels related as 70% and 30% */
#define CELL_SPLIT80x20 0x0800   /* Sizes of panels related as 80% and 20% */
#define CELL_SPLIT90x10 0x0900   /* Sizes of panels related as 90% and 10% */
#define CELL_SPLIT_REL  0x0F00

#define TB_BUBBLE       0x0001  /* Toolbar has bubble help */
#define TB_VERTICAL     0x0002  /* Default toolbar view is vertical */
#define TB_FLOATING     0x0004  /* Toolbar not attached */
#define TB_ATTACHED_LT  0x0010  /* Toolbar attached to left side  */
#define TB_ATTACHED_TP  0x0020  /* Toolbar attached to right side */
#define TB_ATTACHED_RT  0x0040  /* Toolbar attached to top side   */
#define TB_ATTACHED_BT  0x0080  /* Toolbar attached to bottom     */
#define TB_ALLOWED      0x00FF  /* */

#define TB_SEPARATOR    0x7001  /* Separator Item ID */
#define TB_BUBBLEID     0x7002  /* Bubble help window ID */

/* Limits */

#define SPLITBAR_WIDTH       2      /* wodth of split bar between cells */
#define HAND_SIZE            8      /* toolbar drag 'hand' */
#define TB_SEP_SIZE          7      /* width of toolbar separator item */
#define TB_BUBBLE_SIZE      32      /* bubble help item size */
#define CELL_TOP_LIMIT      98      /* maximal space occupied by one cell (%%) */
#define CELL_BOTTOM_LIMIT    2      /* minimal space occupied by one cell (%%) */

/* Window classes */

typedef struct stCellDef
{
    LONG lType;    //Cell type flags
    PSZ  pszClass; //if flag CELL_WINDOW is set, this is a Window Class
    PSZ  pszName;  // Caption
    ULONG ulStyle; //if flag CELL_WINDOW is not set, this a Frame creation flags
    ULONG ulID;    // Cell window ID
    struct stCellDef* pPanel1;
    struct stCellDef* pPanel2;
    PFNWP pClassProc;
    PFNWP pClientClassProc;
    LONG  lSize; //Meaningful only if both CELL_SIZE(1|2) and CELL_FIXED is set
} CellDef;

typedef struct
{
    LONG  lType;   // Toolbar flags
    ULONG ulID;    // Toolbar window ID
    ULONG *tbItems;
} TbDef;

/*
** Internal cell data, used by ordinary windows.
** May be useful for user-defined windows
*/

typedef struct
{
    PFNWP pOldProc;
} WindowCellCtlData;

/* Prototypes */

void ToolkitInit(HAB hab);

HWND CreateCell(CellDef* pCell, HWND hWndParent, HWND hWndOwner);
HWND CellWindowFromID(HWND hwndCell, ULONG ulID);
HWND CellParentWindowFromID(HWND hwndCell, ULONG ulID);
void CreateToolbar(HWND hwndCell, TbDef* pTb);
void GenResIDStr(CHAR *buff, ULONG ulID);

/* Some useful additions */

LONG GetSplit(HWND, LONG lID);
LONG SetSplit(HWND, LONG lID, LONG lProp);
void SetSplitType(HWND, LONG lID, LONG lType);
LONG GetSplitType(HWND, LONG lID);
void ShowCell(HWND hwnd, LONG lID, BOOL bAction);

#endif  /* __CELL_H */

